geom/alg_linear_interpolation.go

```diff
@@ -0,0 +1,57 @@
+package geom
+
+import (
+	"math"
+	"sort"
+)
+
+type linearInterpolator struct {
+	seq        Sequence
+	cumulative []float64
+	total      float64
+}
+
+func newLinearInterpolator(seq Sequence) linearInterpolator {
+	n := seq.Length()
+	if n == 0 {
+		panic("empty seq in newLinearInterpolator")
+	}
+	var total float64
+	cumulative := make([]float64, n-1)
+	for i := 0; i < n-1; i++ {
+		total += seq.GetXY(i).distanceTo(seq.GetXY(i + 1))
+		cumulative[i] = total
+	}
+	return linearInterpolator{seq, cumulative, total}
+}
+
+func (l linearInterpolator) interpolate(frac float64) Point {
+	frac = math.Max(0, math.Min(1, frac))
+	idx := sort.SearchFloat64s(l.cumulative, frac*l.total)
+	if idx == l.seq.Length() {
+		return l.seq.Get(idx - 1).asUncheckedPoint()
+	}
+
+	p0 := l.seq.Get(idx)
+	p1 := l.seq.Get(idx + 1)
+
+	partial := frac * l.total
+	if idx-1 >= 0 {
+		partial -= l.cumulative[idx-1]
+	}
+	partial /= p0.XY.distanceTo(p1.XY)
+
+	return Coordinates{
+		XY: XY{
+			X: lerp(p0.X, p1.X, partial),
+			Y: lerp(p0.Y, p1.Y, partial),
+		},
+		Z:    lerp(p0.Z, p1.Z, partial),
+		M:    lerp(p0.M, p1.M, partial),
+		Type: l.seq.CoordinatesType(),
+	}.asUncheckedPoint()
+}
+
+func lerp(a, b, ratio float64) float64 {
+	return a*(1-ratio) + b*ratio
```

Comment:
I feel the following might be easier to understand:
```
a + ratio*(b - a)
```

At least, my mental model is that, given a line (on the X or Y, etc. plane), compute its length and get the ratio of that length, then add it to the start of the line. It's quite possible my understanding/math is incorrect as well. 😄 

What do you think?

Reply:
That's good feedback. My mental modal (which might be a bit more complicated) is that we're "mixing" `a` and `b` together using a linear combination that sums to `1.0`. So `a` gets `1-ratio` of the mix, and `b` gets `ratio` of the mix. It's a bit confusing though, because the `1-ratio` goes on the _first_ point, rather than the second.

I think your way of formulating it is a bit easier to follow geometrically, I've changed to formulation in e2a08b77e211207bcd44418dc96abddad9431bbd.

Reply:
I ran into this article, which gives some interesting insight into a better way to write a LERP function: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0811r2.html Issue is here: https://github.com/peterstace/simplefeatures/issues/447